From ed3bf8cc300addfba0417a50a02376183a162a88 Mon Sep 17 00:00:00 2001
From: Chilam Ng <chilam@broadcom.com>
Date: Mon, 18 Aug 2014 15:24:11 -0700
Subject: [PATCH 62/77] net: wireless: bcmdhd: fix buffer overrun in link
 statistics

Fixed buffer overrun bug in original patch that caused kernel panic

Bug: 17103277
Change-Id: Ia80f27f148c79801b9800a0ad20472b9231f24b9
Signed-off-by: Chilam Ng <chilam@broadcom.com>
Signed-off-by: Iliyan Malchev <malchev@google.com>
---
 drivers/net/wireless/bcmdhd/wl_cfgvendor.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/bcmdhd/wl_cfgvendor.c b/drivers/net/wireless/bcmdhd/wl_cfgvendor.c
index 2fc983c02ceb..5f854f2c270d 100644
--- a/drivers/net/wireless/bcmdhd/wl_cfgvendor.c
+++ b/drivers/net/wireless/bcmdhd/wl_cfgvendor.c
@@ -1035,16 +1035,18 @@ static int wl_cfgvendor_lstats_get_info(struct wiphy *wiphy,
 	WL_INFORM(("%s: Enter \n", __func__));
 
 	bzero(cfg->ioctl_buf, WLC_IOCTL_MAXLEN);
+	bzero(iovar_buf, WLC_IOCTL_MAXLEN);
 
 	output = cfg->ioctl_buf;
 	radio = (wifi_radio_stat *)output;
 
 	err = wldev_iovar_getbuf(bcmcfg_to_prmry_ndev(cfg), "radiostat", NULL, 0,
-		output, WLC_IOCTL_MAXLEN, NULL);
+		iovar_buf, WLC_IOCTL_MAXLEN, NULL);
 	if (unlikely(err)) {
 		WL_ERR(("error (%d) - size = %zu\n", err, sizeof(wifi_radio_stat)));
 		return err;
 	}
+	memcpy(output, iovar_buf, sizeof(wifi_radio_stat));
 
 	radio->num_channels = NUM_CHAN;
 	output += sizeof(wifi_radio_stat);
@@ -1099,14 +1101,16 @@ static int wl_cfgvendor_lstats_get_info(struct wiphy *wiphy,
 	iface->num_peers = NUM_PEER;
 	iface->peer_info->num_rate = NUM_RATE;
 
+	bzero(iovar_buf, WLC_IOCTL_MAXLEN);
 	output = (char *)iface + sizeof(wifi_iface_stat) + NUM_PEER*sizeof(wifi_peer_info);
 
 	err = wldev_iovar_getbuf(bcmcfg_to_prmry_ndev(cfg), "ratestat", NULL, 0,
-		output, WLC_IOCTL_MAXLEN, NULL);
+		iovar_buf, WLC_IOCTL_MAXLEN, NULL);
 	if (unlikely(err)) {
 		WL_ERR(("error (%d) - size = %zu\n", err, NUM_RATE*sizeof(wifi_rate_stat)));
 		return err;
 	}
+	memcpy(output, iovar_buf, NUM_RATE*sizeof(wifi_rate_stat));
 
 	err =  wl_cfgvendor_send_cmd_reply(wiphy, bcmcfg_to_prmry_ndev(cfg),
 		cfg->ioctl_buf, sizeof(wifi_radio_stat)+NUM_CHAN*sizeof(wifi_channel_stat)+
-- 
2.0.3

